notebooks/1-sequence-stats.jmd

```diff
@@ -0,0 +1,147 @@
+using Plots: make_fillrange_from_ribbon
+# Coronavirus Analysis
+Working directory must be Cov2_Analysis directory
+
+## Mean and Standard Deviation
+Finding mean sequence length, standard devation for length, mean gc content and standard deviation for gc content for CoV sequences.
+
+```julia 
+using BISC195Bioinformatics
+
+return mean_and_std_fasta("data/CoV_sequences.fasta")
+```
+
+## Minimum and Maximum
+
+Finding length of smallest and largest sequence for CoV sequences. 
+
+```julia
+seq_lengths = sequence_lengths("data/CoV_sequences.fasta")
+
+min_length = minimum(seq_lengths)
+max_length = maximum(seq_lengths)
+
+return min_length, max_length
+```
+
+## Identifying Outliers
+
+Using histogram to spot sequence length outliers in CoV sequences. Added more descriptive axis labels and removed legend.
+
+```julia
+using Plots
+
+histogram(seq_lengths, 
+xlabel = "Sequence Length",
+ylabel = "Frequency",
+legend = false)
+```
+## Filtering sequences with more than 25k bases
+
+Modifying sequence_length function from BISC195 Bioinformatics package to output new vectors only containing sequences with >25k bases.
+(Also creates new vector with adjoining headers).
+
+I modified the sequence_length function instead of just calling it because sequence_length already pulls the header and sequence arrays from parse_fasta. 
+This way I don't have to call parse_fasta a second time to get the original vectors for a slice. 
+Included sequence lengths in tuple output for use in next step.
+
+```julia
+"""
+    morethan25k(path)
+
+Given a path to a fasta file, outputs tuple containing only sequences, headers, and sequence lengths for sequences with more than 25k bases.
+"""
+function morethan25k(path)
+    seq_lengths = []
+    headers = parse_fasta(path)[1]
+    sequences = parse_fasta(path)[2]
```

Comment:
Here, you're calling `parse_fasta` twice. When your fasta file is quite large, this can take a lot of extra time. Instead, I'd do `(headers, sequences) = parse_fasta(path)` to do them both in one go. Or, if you find it clearer:

```
headers_sequences = parse_fasta(path)
headers = headers_sequences[1]
sequences = headers_sequences[2]
```

Alternatively, (even better, in my opinion) would be to write this function as

```
function morethan25k(headers, sequences)
#...
end
```

then, in your notebook, do

```
headers, sequences = parse_fasta(path)

seq_length = morethan25k(headers, sequences)
```

The more you can make functions do just one thing, the easier it is to debug or to modify later.

Reply:
Should I move all functions that I wrote for this assignment to my BISC195Bioinformatics package? My morethan25k function is already in my notebook so I wasn't sure what you meant here

Reply:
1. Ultimately, yes, it makes sense to move functions into the package. You can print the docstring out if you want to show it in your notebook with `@doc morethan25k`.
2. The point I was trying to make here though is separate - I'm saying change the function signature to take a pair of headers / sequences as arguments, rather than the path. There may be other things you want to do with the headers / sequences, and if you're starting from the path each time, you'll have to re-read and re-parse the file (which can be slow for large files)

Reply:
Ok I updated the function to take (headers, sequences) and moved it to the BISC195Bioinformatics package. I just wasn't sure what needed to be in the package and what should stay in the notebook. I left this in my notebook: headers, sequences = parse_fasta("data/CoV_sequences.fasta")
seq_lengths = morethan25k(headers, sequences)[3]